### PR TITLE
[typespec-vscode] Add codefix for triple quate indent

### PR DESCRIPTION
**Pull request by mzhongl524**
Fix: https://github.com/microsoft/typespec/issues/4609

Add codefix for `triple-quote-indent` warning ,and meanwhile add test case for it

### CODE REVIEW

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    id: "triple-quote-indent",
+    label: "Format triple-quote-indent",
+    fix: (context) => {
+      const result: CodeFixEdit[] = [];
+
+      const location = getSourceLocation(diagnosticTarget);
+      const startPos = location.pos;
+      const endPos = location.end;
+      const offSet = 3;
+      const splitOrIndentStr = "\r\n";
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:16]: the name is confusing, it's not a indent, right?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    label: "Format triple-quote-indent",
+    fix: (context) => {
+      const result: CodeFixEdit[] = [];
+      const location = getSourceLocation(diagnosticTarget);
+      const startPos = location.pos + 3;
+      const endPos = location.end - 3;
+      const splitStr = "\r\n";
+
+      const lines = splitLines(location.file.text, startPos, endPos);
+
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:17]: handle the special case first, like 0 line, 1 line.

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      });
+      if (minIndentLine.indent <= lastLine.indent) {
+        let indentDiff = lastLine.indent - minIndentLine.indent;
+        const prefix = " ".repeat(indentDiff);
+        // start triple-quote is not on a new line
+        if (firstLine.lineText.trim() !== "") {
+          result.push(context.prependText({ ...location, pos: startPos }, splitStr + prefix));
+        }
+
+        if (lastLine.lineText.trim() === lines[lines.length - 1].lineText.trim()) {
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:39]: what's this for?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          lines.pop();
+        }
+
+        if (lines.length > 0 && firstLine.lineText.trim() === lines[0].lineText.trim()) {
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:43]: what's this for?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        });
+        // end triple-quote is not on a new line
+        if (lastLine.lineText.trim() !== "") {
+          indentDiff =
+            minIndentLine.indent === lastLine.indent
+              ? lastLine.indent
+              : lastLine.indent - minIndentLine.indent;
+          result.push(
+            context.prependText({ ...location, pos: endPos }, splitStr + " ".repeat(indentDiff)),
+          );
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:58]: dont you need a '\n'?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+      const minIndentLine: { startPos: number; indent: number } = lines.reduce((prev, curr) => {
+        return prev.indent < curr.indent ? prev : curr;
+      });
+      if (minIndentLine.indent <= lastLine.indent) {
+        let indentDiff = lastLine.indent - minIndentLine.indent;
+        const prefix = " ".repeat(indentDiff);
+        // start triple-quote is not on a new line
+        if (firstLine.lineText.trim() !== "") {
+          result.push(context.prependText({ ...location, pos: startPos }, splitStr + prefix));
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:36]: instead of adding fix for each line, can't we just have one fix for the whole string?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    const lineText = text.slice(start, pos);
+    const indentNumb = getIndentNumbInLine(lineText);
+    lines.push({ startPos: start + 2, indent: indentNumb, lineText });
+    start = pos;
+  }
+}
+
+function getIndentNumbInLine(lineText: string): number {
+  let curStart = 0;
+  const curEnd = lineText.length;
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:109]: usually end means a position, not lenght

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+          );
+        }
+      }
+
+      return result;
+    },
+  });
+}
+
+function splitLines(
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:67]: any reason not use the existing one?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    label: "Format triple-quote-indent",
+    fix: (context) => {
+      const location = getSourceLocation(diagnosticTarget);
+      const text = location.file.text.slice(location.pos + 3, location.end - 3);
+      const splitStr = "\r\n";
+      const tripleQuote = '"""';
+
+      const lines = splitLines(text);
+      if (lines.length === 0) {
+        return;
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:18]: what would happen if just return?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+      let lastLine = lines[lines.length - 1];
+      if (lastLine.trim() === "") {
+        lines.pop();
+        lastTripleQuoteOnNewLine = true;
+      }
+
+      const minIndentNumb = Math.min(...lines.map((line) => getIndentNumbInLine(line)));
+      const lastLineIndentNumb = getIndentNumbInLine(lastLine);
+      if (minIndentNumb <= lastLineIndentNumb) {
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:46]: do we need to do anything if it's ===?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+          // end triple-quote is not on a new line
+          // this has already been processed, so the removal will not be processed again.
+          lastLine = `${lastLine}${splitStr}${prefix}${tripleQuote}`;
+          lines.pop();
+        }
+
+        // Only indentation is left in the middle
+        const middle = lines
+          .map((line) => {
+            return `${line.slice(0, 2)}${prefix}${line.slice(2, line.length)}`;
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:72]: is the 2 for the newline? what if it's only /n ?

Comment (mzhongl524) [reply packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:72]: yes, `\r` is carriage return, and` \n` is line feed

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts (hunk 9 above)
Comment (mzhongl524) [reply packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:18]: nothing, function returns void

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts (hunk 10 above)
Comment (mzhongl524) [reply packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:46]: I've dealt with the final indentation number if it's ===, immediately after the if statement

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import { defineCodeFix, getSourceLocation } from "../diagnostics.js";
+import type { DiagnosticTarget } from "../types.js";
+
+export function createTripleQuoteIndentCodeFix(diagnosticTarget: DiagnosticTarget) {
+  return defineCodeFix({
+    id: "triple-quote-indent",
+    label: "Format triple-quote-indent",
+    fix: (context) => {
+      const location = getSourceLocation(diagnosticTarget);
+      const text = location.file.text.slice(location.pos + 3, location.end - 3);
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:12]: it's better to use tripleQuote.length than magic number 3

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+          firstLine = `${tripleQuote}${splitStr}${prefix}${firstLine}`;
+          lines.shift();
+        }
+
+        if (lastTripleQuoteOnNewLine) {
+          lastLine = `${lastLine}${tripleQuote}`;
+        } else {
+          // end triple-quote is not on a new line
+          // this has already been processed, so the removal will not be processed again.
+          lastLine = `${lastLine}${splitStr}${prefix}${tripleQuote}`;
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:64]: the indent of the last """ is prefix? shouldn't it be lastLineIndentNumb?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+          lines.pop();
+        }
+
+        // Only indentation is left in the middle
+        const middle = lines
+          .map((line) => {
+            return minIndentNumb !== lastLineIndentNumb
+              ? `${splitStr}${prefix}${line.trim()}`
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:72]: why line.trim()? the pre/post whitespace should be persisted, aren't they?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+function getIndentNumbInLine(lineText: string): number {
+  let curStart = 0;
+  const text = lineText.replace(/(\r\n|\n|\r)/gm, "\r\n");
+  const len = text.length;
+  const flag =
+    len >= 2 &&
+    text.charCodeAt(curStart) === CharCode.CarriageReturn &&
+    text.charCodeAt(curStart + 1) === CharCode.LineFeed;
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:92]: I think the split method has split all the new lines and remove the newline chars, so do we need to consider this again here?

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+import type { DiagnosticTarget } from "../types.js";
+
+export function createTripleQuoteIndentCodeFix(diagnosticTarget: DiagnosticTarget) {
+  return defineCodeFix({
+    id: "triple-quote-indent",
+    label: "Format triple-quote-indent",
+    fix: (context) => {
+      const location = getSourceLocation(diagnosticTarget);
+      const splitStr = "\r\n";
+      const regex = /(\r\n|\n|\r)/gm;
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:13]: please give it a more meaningful name like newlineRegex

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+      const splitStr = "\r\n";
+      const newlineRegex = /(\r\n|\n|\r)/gm;
+      const tripleQuote = '"""';
+      const tripleQuoteLen = tripleQuote.length;
+      const text = location.file.text.slice(
+        location.pos + tripleQuoteLen,
+        location.end - tripleQuoteLen,
+      );
+
+      const lines = splitLines(text);
```
Comment (RodgeFu) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:21]: how about replace the newline chars here so that you dont need to worry about it in other places later?

**Comment by mzhongl524** [comment]
@timotheeguerin  Please help to review this PR, thanks

File: packages/compiler/src/core/scanner.ts [hunk 18]
```diff
@@ -1101,8 +1101,6 @@ export function createScanner(
         end--;
       }
       end--;
-    } else {
```
Comment (timotheeguerin) [inline packages/compiler/src/core/scanner.ts:1104]: why is this error removed

File: packages/compiler/src/core/parser.ts [hunk 19]
```diff
@@ -3427,7 +3429,18 @@ function createParser(code: string | SourceFile, options: ParseOptions = {}): Pa
     if (diagnostic.severity === "error") {
       parseErrorInNextFinishedNode = true;
       treePrintable = false;
+
+      const tripleQuoteWarningCode: string[] = [
```
Comment (timotheeguerin) [inline packages/compiler/src/core/parser.ts:3433]: can those not be added when the diagnostic is emitted instead of having this

File: packages/compiler/test/core/compiler-code-fixes/triple-quote-indent.codefix.test.ts [hunk 20]
```diff
@@ -0,0 +1,77 @@
+import { strictEqual } from "assert";
+import { describe, it } from "vitest";
+import { createTripleQuoteIndentCodeFix } from "../../../src/core/compiler-code-fixes/triple-quote-indent.codefix.js";
+import { SyntaxKind } from "../../../src/index.js";
+import { expectCodeFixOnAst } from "../../../src/testing/code-fix-testing.js";
+
+describe("CodeFix: triple-quote-indent", () => {
```
Comment (timotheeguerin) [inline packages/compiler/test/core/compiler-code-fixes/triple-quote-indent.codefix.test.ts:7]: remove the top level describe, the file is already describing that

File: packages/compiler/test/core/compiler-code-fixes/triple-quote-indent.codefix.test.ts [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+  it("case 1: each triple-quote is on a new line", async () => {
+    await expectCodeFixOnAst(
+      `
+      const a = ┆"""\r\none\r\n  two\r\n  """;
```
Comment (timotheeguerin) [inline packages/compiler/test/core/compiler-code-fixes/triple-quote-indent.codefix.test.ts:11]: please don't use `\r\n` in test, can this not be just done with actual new lines, thjis is hard to read

File: packages/compiler/test/core/compiler-code-fixes/triple-quote-indent.codefix.test.ts [hunk 22]
```diff
@@ -0,0 +1,77 @@
+import { strictEqual } from "assert";
+import { describe, it } from "vitest";
+import { createTripleQuoteIndentCodeFix } from "../../../src/core/compiler-code-fixes/triple-quote-indent.codefix.js";
+import { SyntaxKind } from "../../../src/index.js";
+import { expectCodeFixOnAst } from "../../../src/testing/code-fix-testing.js";
+
+describe("CodeFix: triple-quote-indent", () => {
+  it("case 1: each triple-quote is on a new line", async () => {
```
Comment (timotheeguerin) [inline packages/compiler/test/core/compiler-code-fixes/triple-quote-indent.codefix.test.ts:8]: remove `case x:` in all those tests

File: packages/compiler/src/formatter/print/printer.ts [hunk 23]
```diff
@@ -1920,7 +1920,7 @@ export function printStringTemplateExpression(
   }
 }
 
-function splitLines(text: string): string[] {
+export function splitLines(text: string): string[] {
```
Comment (timotheeguerin) [inline packages/compiler/src/formatter/print/printer.ts:1923]: i think this should be moved out of the printer. Check `syntax-utils` where we moved some other printing related utils.

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+import { defineCodeFix, getSourceLocation } from "../diagnostics.js";
+import type { DiagnosticTarget } from "../types.js";
+
+export function createTripleQuoteIndentCodeFix(diagnosticTarget: DiagnosticTarget) {
+  return defineCodeFix({
+    id: "triple-quote-indent",
+    label: "Format triple-quote-indent",
+    fix: (context) => {
+      const location = getSourceLocation(diagnosticTarget);
+      const splitStr = "\r\n";
```
[the remark this answers is not in the document]
Comment (timotheeguerin) [reply packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:12]: the formatter wil reformat to windows if needs be, but by default we always do everything in unix style in typespec

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+      const newlineRegex = /(\r\n|\n|\r)/gm;
+      const tripleQuote = '"""';
+      const tripleQuoteLen = tripleQuote.length;
+      const text = location.file.text.slice(
+        location.pos + tripleQuoteLen,
+        location.end - tripleQuoteLen,
+      );
+
+      const lines = splitLines(text).map((line) => line.replace(newlineRegex, ""));
```
Comment (timotheeguerin) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:21]: kinda feels like there is a bug in splitLines, it shouldn't include the new lines and have you remove them again.

File: packages/compiler/src/formatter/print/printer.ts (hunk 23 above)
Comment (timotheeguerin) [reply packages/compiler/src/formatter/print/printer.ts:1923]: as the other comment also found, seems like there is a bug, it shouldn't include the new lines. I think it kinda works for the formatter use because it counts the newline in the `whitespaceIndent` but if you fix the function it work as well.

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+
+      const middle = lines.map((line) => `${splitStr}${prefix}${line}`).join("");
+      return context.replaceText(
+        location,
+        `${tripleQuote}${middle}${splitStr}${" ".repeat(lastLineIndentNumb)}${tripleQuote}`,
+      );
+
+      function getIndentNumbInLine(lineText: string): number {
+        let curStart = 0;
+        const text = lineText.replace(newlineRegex, "");
```
Comment (timotheeguerin) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:60]: I don't get when this happens that there is still new lines left. Can you explain. 
tldr that `newlineRegex` should be removed and try not to depend on those.

File: packages/compiler/src/core/scanner.ts (hunk 18 above)
Comment (mzhongl524) [reply packages/compiler/src/core/scanner.ts:1104]: This has been repeated in the following `findTripleQuotedStringIndent` method

File: packages/compiler/src/core/parser.ts (hunk 19 above)
Comment (mzhongl524) [reply packages/compiler/src/core/parser.ts:3433]: After testing, it makes the most reasonable to add the generated codefix here, if the codefix generated in the program.ts file or elsewhere is too repetitive, I don't know if there is any good way to solve it? or something better?

Comment (mzhongl524) [reply packages/compiler/src/core/parser.ts:3433]: the code has been updated and submitted based on other comments

Comment (timotheeguerin) [reply packages/compiler/src/core/parser.ts:3433]: this is just that you are injecting those into a place that is not where they were produced(mutating something else)

what is the issue with doing this in the scanner
```ts
 error({
          code: "triple-quote-indent",
          codeFixes: [createTripleQuoteIndentFix({ file, pos, end })],
        });
```

this also makes it more obvious for the diagnostic target and you should be able to change the type to directly be a location 
instead of having to call `getSourceLocation(diagnosticTarget);` which is opaque and in theory might not necessarly target what you want.

File: .chronus/changes/add-codefix-for-triple-quate-indent-2025-1-6-8-10-18.md [hunk 27]
```diff
@@ -0,0 +1,8 @@
+---
+# Change versionKind to one of: internal, fix, dependencies, feature, deprecation, breaking
+changeKind: feature
+packages:
+  - "@typespec/compiler"
+---
+
+Add codefix for triple-quote value formatting, currently supported warning 'triple-quote-indent','no-new-line-end-triple-quote' and 'no-new-line-start-triple-quote'
```
Comment (timotheeguerin) [inline .chronus/changes/add-codefix-for-triple-quate-indent-2025-1-6-8-10-18.md:8]:
```suggestion
Add codefix for triple-quote value formatting, currently supported warning `triple-quote-indent`, `no-new-line-end-triple-quote` and `no-new-line-start-triple-quote`
```

File: packages/compiler/src/core/parser.ts (hunk 19 above)
Comment (mzhongl524) [reply packages/compiler/src/core/parser.ts]: the parameter type is a union type that contains the location, and in the `getSourceLocation(diagnosticTarget)` method, if it is a location, it will be returned directly.
 readjust the code with the new comment and commit

File: packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts [hunk 28]
```diff
@@ -0,0 +1,69 @@
+import { isWhiteSpaceSingleLine } from "../charcode.js";
+import { defineCodeFix, getSourceLocation } from "../diagnostics.js";
+import { splitLines } from "../helpers/syntax-utils.js";
+import type { DiagnosticTarget } from "../types.js";
+
+export function createTripleQuoteIndentCodeFix(diagnosticTarget: DiagnosticTarget) {
```
Comment (timotheeguerin) [inline packages/compiler/src/core/compiler-code-fixes/triple-quote-indent.codefix.ts:6]: the point of changing where you called was  this could also be changed to `SourceLocation` and remove the call to `getSourceLocation` inside.

**Review by timotheeguerin — APPROVED** [review]
